File: v6/prolog/labs/rel_spreading/probes/ts/c3_row_spread_position.ts
// C3 row spread in a head position: can a spread be followed by explicit
// positional arguments, and where does the width come from?

type ARow = [id: number, name: string];

// (1) a spread with trailing explicit elements in a TYPE position: accepted
type CRow = [...ARow, extra: number];
declare const aRow: ARow;
const cFromSpread: CRow = [...aRow, 5];

// (2) the same shape in a CALL: accepted only because the spread argument has
// a tuple type. The width is read from the tuple type, not from the value.
declare function takesThree(id: number, name: string, extra: number): void;
takesThree(...aRow, 5);

// (3) an ARRAY (unknown width) in the same position: refused
declare const aArray: number[];
declare function takesTwoNumbers(x: number, y: number): void;
takesTwoNumbers(...aArray);

// (4) each spliced position is an independent slot, not one aggregate value
const idSlot: number = cFromSpread[0];
const nameSlot: string = cFromSpread[1];

export { cFromSpread, idSlot, nameSlot };

File: v6/prolog/labs/rel_spreading/probes/rust/c3_row_spread_position.rs
// C3 row spread in a call/head position: can a Rust tuple be spliced into an
// argument list with trailing explicit arguments?

fn takes_three(_id: i64, _name: &str, _extra: i64) {}

fn main() {
    let a: (i64, &str) = (1, "n");

    // (1) splice the tuple then append one argument: no such syntax
    takes_three(..a, 5);

    // (2) the tuple itself is one value, not three slots
    takes_three(a, 5);

    // (3) the only spread-shaped thing is `..` in a PATTERN, which discards
    // rather than binds
    let (id, ..) = (1i64, "n", 5i64);
    println!("{}", id);
}
